ships/0021-local-source-upload.md

```diff
@@ -0,0 +1,348 @@
+with other applications, and producing a stand alone container-image. The image must be based on
+UBI8, following the same standards than others.
+
+This image also has `tar` installed, which will be used for the data upload mechanism, as explained
+before.
+
+Another requirement is following the same configuration standard than other external images used by
+the Build Controller, like a environment variable (i.e.: `GIT_CONTAINER_TEMPLATE`) to overwrite the
+image before the controller starts.
+
+At last, it should provide meaningful information for Tekton's `TaskResult`, so we can differentiate
+this step run from others. For the `waiter` process results, it can take advantage of the
+[termination-log pattern][KubernetesTerminationLog] to inform Tekton more detailed execution results.
+
+### Use-Cases
+
+#### Developer's Inner-Loop
+
+As a software developer, I'm able to try out the local changes being worked on with a container-image
+built by the same Shipwright Build-Controller and using cluster's computing power. Another positive
+side effect is that developers won't need to install a container-manager, like for instance Docker,
+to be able to build new images.
+
+#### Continuous-Integration
+
+During CI pipelines Shipwright Build-Controller can be employed to build container-images based on
+the local repository clone. That can help teams to assert their container-image build process works,
+and also create new images for more complex test scenarios.
+
+Additionally, it would also avoid the need for a container-manager (like Docker) to build
+container-images during pipelines, so it also simplifies the process of testing applications.
+
+### Test Plan
+
+Testing this feature can be achieved by unit-testing on each component. The CLI can simulate the
+works of the Build Controller, as in simululate how the Kubernetes API will respond upon each
+situation.
+
+The Build Controller will also simulate Kubernetes API during the unit-testing, and for end-to-end
+testing we will perform data upload, as the CLI does, in order to have the testing complete.
+
+### Risks and Mitigations
+
+In a nutshell, this enhancement-proposal will use `exec` to run `tar` on the waiter container, and
+stream the data via standard-input, the data lies in a `EmptyDir` volume.
+
+#### Abuse of `exec`
+
+Via the CLI users should only be able to execute the `tar` command inside, so Waiter's base-image
+must be rid of additional tooling. Additonally, waiter should run as non-privileged container, with a
+non-root user.
+
+#### Excessive Amounts of Data
+
+Users could potentially upload more data than the node running the actual build is able to support.
+Here we need to make sure the CLI only uploads intersting bits, like for instance ignoring the
+entries on `.gitignore` (and maybe other types of ignore-files), plus we can enhance the Waiter to
+chech how much data has been uploaded and enforce limits.
```

Comment:
> plus we can enhance the Waiter to chech how much data has been uploaded and enforce limits.

This might not be necessary. You wrote that the container template can be defined. This can define a resource limit for the ephemeral storage. This might already do it. Though, I am always not sure how ephemeral storage limits defined individually on containers in a pod work with an emptyDir that is shared by more than one container in this pod.

(And for Git we have no other means as well.)

Reply:
Sure! I don't know the details of ephemeral storage in k8s yet, so it came along as a suggestion, something that we may need during the implementation phase.

Reply:
there is a beta level feature for limiting ephemeral storage where you enable a feature gate 

in the doc I see, you can do this at the container level, and then things get added up for the pod level

https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage
https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#configurations-for-local-ephemeral-storage
https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#setting-requests-and-limits-for-local-ephemeral-storage